src/log_handling/RetailLogHandler.ts

```
@@ -0,0 +1,344 @@
+
+        if (!this.activity) {
+            console.error("[RetailLogHandler] Challenge mode stop with no active ChallengeModeDungeon");
+            return;
+        }
+
+        const challengeModeActivity = this.activity as ChallengeModeDungeon;
+        const endDate = line.date();
+
+        // The actual log duration of the dungeon, from which keystone upgrade
+        // levels can be calculated. This includes player death penalty. 
+        const CMDuration = Math.round(parseInt(line.arg(4), 10) / 1000);
+
+        challengeModeActivity.endChallengeMode(endDate, CMDuration);
+        this.endRecording(this.activity);
+    };
+
+    handleEncounterStartLine(line: LogLine) {
+        console.debug("[RetailLogHandler] Handling ENCOUNTER_START line:", line);
+
+        if (!this.activity) {
+            super.handleEncounterStartLine(line);
+        } else {
+            const category = this.activity.getCategory();
+            const isChallengeMode = category === VideoCategory.MythicPlus;
+
+            if (isChallengeMode) {
+                const activeChallengeMode = this.activity as ChallengeModeDungeon;
+                const encounterID = parseInt(line.arg(1), 10)
+                const eventDate = line.date();
+                
+                const segment = new ChallengeModeTimelineSegment(
+                    TimelineSegmentType.BossEncounter,
+                    eventDate,
+                    this.getRelativeTimestampForTimelineSegment(eventDate),
+                    encounterID
+                );
+    
+                activeChallengeMode.addTimelineSegment(segment, eventDate);
+                console.debug(`[RetailLogHandler] Starting new boss encounter: ${dungeonEncounters[encounterID]}`)
+            }
+        }
+    }
+
+    handleEncounterEndLine(line: LogLine) {
+        console.debug("[RetailLogHandler] Handling ENCOUNTER_END line:", line);
+
+        if (this.activity) {
```

Same as above, return and fail early and often.